File: Python/skebbyAPI.py
#!/usr/bin/python

# Skebby Python Gateway Class
# Importing useful libraries
import urllib
import urllib2

# define constants
NET_ERROR = "Network error, unable to send the message"
SENDER_ERROR = "You can specify only one type of sender, numeric or alphanumeric"
url = 'http://api.skebby.it/api/send/smseasy/advanced/http.php'

# The actual class
class skebbyAPI:

    # init
    def __init__(self, usr, pwd):
        self.usr = usr
        self.pwd = pwd
        self.charset = 'UTF-8'

    # useful is_array function
    def _isArray(self,obj): 
        is_array = lambda var: isinstance(var, (list, tuple))
        return is_array(obj)

    # urlEncode function for recipients
    def _urlEncode(self,recipients):
        resultString = ''
        if isinstance(recipients[0],dict) and ('recipient' in recipients[0]):
            for i in range(len(recipients)):
                for key,value in recipients[i].items():
                    resultString = resultString + 'recipients[' + str(i) + '][' + str(key) + ']=' + urllib.quote_plus(value) + '&'
        else:
            if self._isArray(recipients) is False: recipients = [recipients]
            for number in recipients:
                resultString = resultString + 'recipients[]=' + urllib.quote_plus(number) + '&'
        return resultString[:-1]

    # Core function where the class connects to Skebby Gateway URL and sends the data.
    def do_post_request(self,input):
        input['username'] = self.usr
        input['password'] = self.pwd
        rec = ''
        if 'recipients' in input.keys():
            rec = '&' + self._urlEncode(input['recipients'])
            print rec
            #input.pop('recipients',None)

        data = urllib.urlencode(input) + rec
        headers = { 'User-Agent' : 'Skebby Python Gateway' }

        req = urllib2.Request(url, data, headers)
        try:
            response = urllib2.urlopen(req)
        except urllib2.HTTPError as e:
            result = {}
            result['status'] = 'failed'
            result['code'] = e.code
            result['message'] = NET_ERROR
            return result
        except urllib2.URLError as e:
            result = {}
            result['status'] = 'failed'
            result['message'] = e.reason
            return result

        resultString = response.read()
        return resultString

    # send SMS function
    def sendSMS(self,input):
        if 'method' not in input.keys():
            input['method'] = 'send_sms_classic'
        if ( 'sender_number' in input.keys() and 'sender_string' in input.keys() ) and (input['sender_number'] != '' and input['sender_string'] != '') :
            result = {}
            result['status'] = 'failed'
            result['message'] = SENDER_ERROR
            return result

        if 'charset' in input.keys() and input['charset'] != 'ISO-8859-1' :
            input['charset'] = self.charset
        elif 'charset' in input.keys() and input['charset'] == 'ISO-8859-1':
            self.charset = 'ISO-8859-1'
        
        result = self.do_post_request(input)
        return result

    # get Credit function
    def getCredit(self,input = {}):
        input['method'] = 'get_credit'
        if 'charset' in input.keys() and input['charset'] != 'ISO-8859-1' :
            input['charset'] = self.charset
        elif 'charset' in input.keys() and input['charset'] == 'ISO-8859-1':
            self.charset = 'ISO-8859-1'
        
        result = self.do_post_request(input)
        return result

    # Add Alias function
    def addAlias(self,input):
        input['method'] = 'add_alias'
        if 'charset' in input.keys() and input['charset'] != 'ISO-8859-1' :
            input['charset'] = self.charset
        elif 'charset' in input.keys() and input['charset'] == 'ISO-8859-1':
            self.charset = 'ISO-8859-1'
        
        result = self.do_post_request(input)
        return result
        
    def arrayResult(self,resultString):
        results = resultString.split('&')
        result = {}
        for r in results:
            temp = r.split('=')
            result[temp[0]] = urllib.unquote(temp[1]).replace('+',' ')

        return result
    
    def printResult(self,result):
        r = self.arrayResult(result)
        for key,value in r.items():
            print key + ': ' + value

File: Ruby/skebbyAPI.rb
#!/usr/bin/ruby

# Skebby Ruby Gateway Class
# Importing useful libraries
require 'net/http'
require 'uri'
require 'cgi'

# The actual class
class SkebbyAPI
	def initialize(username,password)
		@url = 'http://api.skebby.it/api/send/smseasy/advanced/http.php'
		@parameters = {
			'username' => username,
			'password' => password,
		}
	end
	
	# Core function where the class connects to Skebby Gateway URL and sends the data.
	def do_post_request(data)
		response = Net::HTTP.post_form(URI(@url),data)
        if response.message == "OK"
			result = {}
			response.body.split('&').each do |res|
				if res != ''
					temp = res.split('=')
					if temp.size > 1
						result[temp[0]] = temp[1]
					end
				end
			end
        else
            result = "Error in the HTTP request"
        end
        return result
	end
	
	# Send SMS function
	def sendSMS(recipients,text,method='send_sms_classic',options = {})
		unless recipients.kind_of?(Array)
        	recipients = [recipients]
		end
		
		if (recipients[0].kind_of?(Hash) and recipients[0].has_key?('recipient'))
			for i in 0..(recipients.length-1)
				recipients[i].each do |key,value|
					@parameters["recipients[#{i}][#{key}]"] = value
				end
			end
		else
			@parameters["recipients[]"] = recipients
		end
		
      	@parameters['text'] = text
		@parameters['method'] = method

		unless options[:sender_number].nil?
			@parameters['sender_number'] = options[:sender_number]
		end
	
		unless options[:sender_string].nil?
			@parameters['sender_string'] = options[:sender_string]
		end
	
		unless options[:charset].nil?
			@parameters['charset'] = options[:charset]
		end
	
		unless options[:delivery_start].nil?
			@parameters['delivery_start'] = options[:delivery_start]
		end
	
		unless options[:encoding_scheme].nil?
			@parameters['encoding_scheme'] = options[:encoding_scheme]
		end
	
		unless options[:validity_period].nil?
			@parameters['validity_period'] = options[:validity_period]
		end
	
		unless options[:user_reference].nil?
			@parameters['user_reference'] = options[:user_reference]
		end
		
		result = self.do_post_request(@parameters)
		return result
	end
	
	# Get Credit function
	def getCredit()
		@parameters['method'] = 'get_credit'
		result = self.do_post_request(@parameters)
		return result
	end
	
	# Add Alias function
	def addAlias(alias_name,business_name,nation,vat_number,taxpayer_number,street,city,postcode,contact,charset=nil)
		@parameters['method'] = 'add_alias'
		@parameters['alias'] = alias_name
		@parameters['business_name'] = business_name
		@parameters['nation'] = nation
		@parameters['vat_number'] = vat_number
		@parameters['taxpayer_number'] = taxpayer_number
		@parameters['street'] = street
		@parameters['city'] = city
		@parameters['postcode'] = postcode
		@parameters['contact'] = contact
	
		unless charset.nil?
			@parameters['charset'] = charset
		end
		
		result = self.do_post_request(@parameters)
		return result
	end
	
	# Print result in a fancy way
	def printResult(result)
		if result.has_key?('status') and result['status'] == 'success'
			puts "Success, response contains:"
			result.each do |key,value|
				puts "\t#{key} => #{CGI::unescape(value)}"
			end
			true
		else      
			puts "Error, trace is:"
			result.each do |key,value|
				puts "\t#{key} => #{CGI::unescape(value)}"
			end
			false
		end
	end
end